datalad_next/utils/shell_connection.py

```diff
@@ -0,0 +1,305 @@
+            subprocess_inputs.put(None)
+
+
+class ShellCommandExecutor:
+    """Execute a command in a shell and read its result and return code
+
+    Objects of this class use interactive subprocess channels to:
+     - execute a single POSIX pipeline-list
+     - collect its output
+     - collect its return code
+
+    The command is extended by a postfix that prints a most likely unique
+    end marker and the return code of the command. The end marker is used to
+    detect the end of the output and the determine the return code of the
+    executed command.
+
+    Both, command output and command return code are available via
+    :class:`ShellCommandResponseGenerator`-instances that are returned by the
+    :meth:`execute_command`-method of this class.
+    """
+
+    def __init__(self,
+                 process_inputs: Queue,
+                 stdout: OutputFrom,
+                 ) -> None:
+        self.process_inputs = process_inputs
+        self.stdout = stdout
+        self.end_marker, self.command_postfix = self._get_marker()
+
+    def execute_command(self,
+                        command: bytes,
+                        *,
+                        stdin: Iterable[bytes] | None = None,
+                        ) -> ShellCommandResponseGenerator:
+        """Execute a command in the connected shell
+
+        Parameters
+        ----------
+        command : bytes
+            The command to execute. Command can be a single POSIX pipeline list.
+            TODO: provide an example
+            (Internally the command is extended to mark the end of the output
+            by printing an end marker and the return code.)
+        stdin : Iterable[byte] | None, optional, default: None
+            If given, the bytes are sent to the stdin of the command.
+
+            If the
+            command reads its ``stdin`` until EOF, you have to use
+            :meth:`self.close` to close ``stdin`` of the command. Otherwise
+            the command will usually not terminate.
+            Once :meth:`self.close` is called, no more commands can be executed
+            with this :class:`ShellCommandExecutor`-instance.
+
+        Yields
+        ------
+        :class:`ShellCommandResponseGenerator`
+
+            A generator that yields the output of ``stdout`` of the command.
+            The generator is exhausted when all output is read. After that,
+            the return code of the command execution is available in the
+            ``code``-attribute of the generator.
+        """
+        self.process_inputs.put([command + self.command_postfix])
+        if stdin is not None:
+            self.process_inputs.put(stdin)
+        return ShellCommandResponseGenerator(self.stdout, self.end_marker)
+
+    def upload(self,
+               local_path: Path,
+               remote_path: PurePosixPath,
+               ) -> tuple[int, bytes]:
+        """Upload a local file to a named file in the connected shell
```

Comment:
Should say what the requirement for the upload target are: Must be a POSIX system and must have `dd`.

Reply:
The requiremets are new POSIX system, `head`, and `base64`. All documented in the doc-string